Variation a

DamonNomadJr opened the pull request:
This code now is fully functional for 'Card Service.'
The following files have been updated:
1. src/model/card-form.ts
2. src/model/card.ts
3. src/model/section.ts
4. src/model/tag.ts
5. src/services/tag.service.ts
6. src/services/view.service.ts
This page now can attach tags to elements when created, and after it is created. Furthermore, the tag drop-down list is updated as new tags are added.

DamonNomadJr committed:
Creats card service interface for
the application. It does not contain
any styling.

saaniaki committed:
Merge pull request #1 from DamonNomadJr/Variation-A

Variation A has been accepted.

DamonNomadJr committed:
Fixes the card and tag service and forms

DamonNomadJr committed:
Card model contains specified tags, also on creation

saaniaki commented:
Good job!
There are a couple of small bugs:
* Removing a tag from the tag list does not remove the already added tags of existing tags.
* The initial tags (i.e. t1, t2, t3 and t4) are not being shown in the tag section.
* Adding a new tag doesn't close the form.
* Adding a duplicate tag should not be allowed.

I'm also going to add some comments on the code, please review and do another pull request.

saaniaki committed:
Merge branch 'master' into Variation-A

saaniaki commented on src/services/tag.service.ts:
```diff
@@ -6,24 +6,48 @@ import { Injectable } from "../utility/service-gen";
+    }
+
+    public update(){
+        this.resources.forEach((e) => {
```
Please consider using a for-each loop:
for(const e of this.resources) {
    ....
}

DamonNomadJr replied:
Has been update

saaniaki commented on src/services/tag.service.ts:
```diff
@@ -6,24 +6,48 @@ import { Injectable } from "../utility/service-gen";
 
-    tags: Tag[] = [];
+    tags: Tag[];
+    resources: Function[];
```
Please use a better name, ex. updateProcedures

DamonNomadJr replied:
Renamed resources to updateAgents
and register also was changed to addAgent

saaniaki commented on src/services/tag.service.ts:
```diff
@@ -6,24 +6,48 @@ import { Injectable } from "../utility/service-gen";
 
-    tags: Tag[] = [];
+    tags: Tag[];
+    resources: Function[];
```
This property must be read-only and private

DamonNomadJr replied:
tags are now private and read only.

saaniaki commented on src/services/tag.service.ts:
```diff
@@ -6,24 +6,48 @@ import { Injectable } from "../utility/service-gen";
 export class TagService extends Service {
 
-    tags: Tag[] = [];
+    tags: Tag[];
```
This property must be read-only and private

saaniaki commented on src/services/tag.service.ts:
```diff
@@ -6,24 +6,48 @@ import { Injectable } from "../utility/service-gen";
-    tags: Tag[] = [];
+    tags: Tag[];
+    resources: Function[];
     section: Section;
```
This property is declared but never used.

DamonNomadJr replied:
It is removed.

saaniaki commented on src/model/card-form.ts:
```diff
@@ -1,35 +1,61 @@
 
-    _input: HTMLInputElement;
+    @Inject(TagService) tagService : TagService;
+    _title: HTMLInputElement;
```
I guess you forgot the "private" keyword.

DamonNomadJr replied:
Both are now private

saaniaki commented on src/model/card-form.ts:
```diff
@@ -1,35 +1,61 @@
-    _input: HTMLInputElement;
+    @Inject(TagService) tagService : TagService;
+    _title: HTMLInputElement;
+    _body: HTMLInputElement;
```
I guess you forgot the "private" keyword.

saaniaki reviewed (commented):
Please refactor the code and also remove extra new lines.

DamonNomadJr committed:
Fixes minor code issues and syntaxes

DamonNomadJr committed:
Fixes tag removal sync from both tags and card's tags

DamonNomadJr committed:
Adds resetDom to insertable to recreate DOM if necessary

DamonNomadJr commented:
I have fixed the problem with tag removal. I had to force the commit over the previous version because I had logs in my code. 
I have also added resetDom() to be able to reset the DOM elements so that when creating a form the fields are set empty. I have done this practice for card-form only. resetDom() for card, tag, and tag-form are not set.

I have modified the following files:
```
modified:   src/model/card.ts
modified:   src/model/card-form.ts
modified:   src/model/tag.ts
modified:   src/model/section.ts
modified:   src/model/insertable.ts
modified:   src/model/tag-form.ts
modified:   src/services/tag.service.ts
```
I haven't done the following parts
* The initial tags (i.e. t1, t2, t3 and t4) are not being shown in the tag section.
* Adding a duplicate tag should not be allowed.

saaniaki commented on src/model/card.ts:
```diff
@@ -1,47 +1,177 @@
+        tag_item.className = 'tag-item';
+        tag_item.onclick = () => this.removeTag(t, tag_item);
+        tag_item.textContent = t.name;
+        console.log(t);
```
Please remove the log statement.

saaniaki commented on src/model/card.ts:
```diff
@@ -1,47 +1,177 @@
     protected onInit(...args: any[]): void {
-        this.topic = args[0];
-        this. body = args[1];
+        this.topic = args[0][0];
```
Why this is two dimensional?! pay attention that you're always accessing [0][x]

saaniaki commented on src/model/card.ts:
```diff
@@ -1,47 +1,177 @@
+     * By calling this, it will associate a specific tag with
+     * this object.
+     */
+    private addTag() {
```
what is the difference of this and setTag?! and why you have repeated yourself exactly two lines after?
As a programmer, do you really think you should repeat your code?!

saaniaki commented on src/model/card.ts:
```diff
@@ -1,47 +1,177 @@
+     * Checks if tag exists in Tag Service or no
+     * @param t 
+     */
+    private findTag(t: Tag): Boolean{
```
This method sucks, it can be done like this:

    private findTag(t: Tag): Boolean {
        for (const e of this.tagService.tags)
            if (e == t)
                return false;
        return true;
    }

saaniaki commented on src/model/card.ts:
```diff
@@ -1,47 +1,177 @@
+     * @param t 
+     * @param e 
+     */
+    private removeTag(t: Tag, e: HTMLElement){
```
This method can be simplified

saaniaki commented on src/model/card.ts:
```diff
@@ -1,47 +1,177 @@
+     * the <span> dom under the <div class="tag-Section">.
+     * @param t 
+     */
+    private publishTag(t: Tag){
```
instead of tracking the tag and the parent element consider using encapsulation, i.e. nested class or a separated private class.

saaniaki commented on src/model/card.ts:
```diff
@@ -1,47 +1,177 @@
+            e.parentNode.removeChild(e);
+        }
+        else
+            throw ("Tag Not Found.");
```
There is a huge bug in the program:
1. Create a new tag in the tag section.
2. Create a new card and add the newly created tag to it.
3. Now remove the tag which you created in step 1 from the tag section.
4. Try to create a new card.

You will not be able to create a new card and an exception will be thrown at you:
uncaught exception: Tag Not Found.

saaniaki commented on src/services/tag.service.ts:
```diff
@@ -1,29 +1,55 @@
+        return this._tags;
+    }
+
+    public valueSelect(name: string): Tag{
```
What does this method do? Please either add comments or use a more descriptive name.

saaniaki reviewed (changes requested):
Card class has a lot of problems, it can be simplified and also introduces a bug.
Please consider fixing these and improving the code style.

saaniaki commented on src/services/card.service.ts:
```diff
@@ -6,11 +6,12 @@ import {Section} from "../model/section";
 export class CardService extends Service {
 
-    cards: Card[] = [];
+    cards: Card[];
```
Please make this private and read-only

saaniaki commented on src/services/card.service.ts:
```diff
@@ -6,11 +6,12 @@ import {Section} from "../model/section";
 
-    cards: Card[] = [];
+    cards: Card[];
     section: Section;
```
imported and declread but never used

saaniaki reviewed (changes requested):
Please fix these as well

saaniaki commented:
Overall please consider moving a bigger portion of the tag syncing logic into the services and take advantage of encapsulation.

saaniaki commented:
Please fetch the upstream as I just pushed a new commit and re-based.
